src/useLocalStorage.ts

```diff
@@ -1,22 +1,37 @@
 import { useEffect, useState } from 'react';
 import { isClient } from './util';
 
-type Dispatch<A> = (value: A) => void;
-type SetStateAction<S> = S | ((prevState: S) => S);
+type parserOptions<T> =
+  | {
+      raw: true;
+    }
+  | {
+      raw: false;
+      serializer: (value: T) => string;
+      deserializer: (value: string) => T;
+    };
 
-const useLocalStorage = <T>(key: string, initialValue?: T, raw?: boolean): [T, Dispatch<SetStateAction<T>>] => {
+const useLocalStorage = <T>(
+  key: string,
+  initialValue?: T,
+  options?: parserOptions<T>
+): [T, React.Dispatch<React.SetStateAction<T>>] => {
   if (!isClient) {
     return [initialValue as T, () => {}];
   }
 
+  // Use provided serializer/deserializer or the default ones
+  const serializer = options ? (options.raw ? String : options.serializer) : JSON.stringify;
+  const deserializer = options ? (options.raw ? String : options.deserializer) : JSON.parse;
+
   const [state, setState] = useState<T>(() => {
     try {
       const localStorageValue = localStorage.getItem(key);
-      if (typeof localStorageValue !== 'string') {
-        localStorage.setItem(key, raw ? String(initialValue) : JSON.stringify(initialValue));
-        return initialValue;
+      if (localStorageValue) {
```

Comment:
This probably should check for `undefined`, in case `localStorageValue` is an empty string.

```suggestion
      if (localStorageValue !== undefined) {
```

For example

```js
useLocalStorage('foo', '');
```

Reply:
Woops just missed that. On that note, it should check for `null` instead of `undefined` since `localStorage.getItem()` returns with a type of `string | null`